src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+use crate::edn::{Edn, Error, List, Map, Set, Vector};
+
+pub(crate) fn tokenize(edn: &str) -> std::str::Chars {
+    edn.chars()
+}
+
+pub(crate) fn parse(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    Ok(match c {
+        Some('[') => read_vec(chars)?,
+        Some('(') => read_list(chars)?,
+        Some('@') => read_set(chars)?,
+        Some('{') => read_map(chars)?,
+        edn => parse_edn(edn, chars)?,
+    })
+}
+
+pub(crate) fn parse_edn(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    match c {
+        Some('\"') => Ok(read_str(chars)),
+        Some(':') => Ok(read_key(chars)),
+        Some(n) if n.is_numeric() => Ok(read_number(n, chars)?),
```

Comment:
Have you tried putting this match arm one before the error one? I tend to think ifs inside match clauses are slower than pure pattern matches, I believe Rust optimize those somehow (again, might be wrong).